content/umdctf_2024/tcache.md

```diff
@@ -0,0 +1,241 @@
+---
+title: Tcache 
+date: 2024-04-17T14:23:28+03:00
+description: Information about tcache 
+author: PineBel
+tags:
+    - pwn
+
+----
+## Tcache
+
+Tcache (per thread cache) was added in glibc 2.26 to make heap allocation more efficient, and basically each thread will get it s own bin of small chunks. 
+
+Important things to know:
+
+- Each thread has 64 singly-linked tcache bins (`TCACHE_MAX_BINS`).
+- Each bin has a maximum of 7 chunks of the same size.
+- A bin can have chunks ranging from 24 to 1032 bytes (on a 64-bit system).
+- Each bin only has an `fd` pointer.
+- Allocating from a tcache bin takes priority over every other bin.
+- When `free` is called or `malloc` is invoked, this is the lifecycle (which bins will be used):
+  - Initially checks tcache. If full or the size exceeds limits, looks in fastbins. If conditions are met, searches in unsorted bins, then small or large bins.
+  - For `malloc` only, if no suitable bin is found, it tries obtaining memory from the top chunk. If that fails, it extends the heap. If both attempts fail, it returns null.
+
+<br>
+
+## Tcache Structure
+To understand better how the tcache works, I played a bit around with [shellphishes](https://github.com/shellphish/how2heap) GitHub project. <br>
```

Comment:
you can actually just use two spaces to force a newline in markdown, like <space><space> at the end of the line